packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -742,178 +738,849 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
+    if (selection.extentOffset >= text.length) {
+      return selection;
+    }
+    final int nextExtent = nextCharacter(selection.extentOffset, text, includeWhitespace);
+    return selection.copyWith(extentOffset: nextExtent);
+  }
+
+  // Extend the current selection to the end of the field.
+  //
+  // See also:
+  //   * [_extendSelectionToStart]
+  void _extendSelectionToEnd(SelectionChangedCause cause) {
+    if (selection!.extentOffset == _plainText.length) {
+      return;
+    }
+    if (!selectionEnabled) {
+      return moveSelectionToEnd(cause);
+    }
+
+    final TextSelection nextSelection = selection!.copyWith(
+      extentOffset: _plainText.length,
+    );
+    _updateSelection(nextSelection, cause);
+  }
+
+  // Extend the current selection to the start of the field.
+  //
+  // See also:
+  //   * [_expandSelectionToEnd]
+  void _extendSelectionToStart(SelectionChangedCause cause) {
+    if (selection!.extentOffset == 0) {
       return;
     }
+    if (!selectionEnabled) {
+      return moveSelectionToStart(cause);
+    }
+
+    final TextSelection nextSelection = selection!.copyWith(
+      extentOffset: 0,
+    );
+    _updateSelection(nextSelection, cause);
+  }
+
+  // Returns the TextPosition above or below the given offset.
+  TextPosition _getTextPositionVertical(int textOffset, double verticalOffset) {
+    final Offset caretOffset = _textPainter.getOffsetForCaret(TextPosition(offset: textOffset), _caretPrototype);
+    final Offset caretOffsetTranslated = caretOffset.translate(0.0, verticalOffset);
+    return _textPainter.getPositionForOffset(caretOffsetTranslated);
+  }
+
+  // Returns the TextPosition above the given offset into _plainText.
+  //
+  // If the offset is already on the first line, the given offset will be
+  // returned.
+  TextPosition _getTextPositionAbove(int offset) {
+    // The caret offset gives a location in the upper left hand corner of
+    // the caret so the middle of the line above is a half line above that
+    // point and the line below is 1.5 lines below that point.
+    final double preferredLineHeight = _textPainter.preferredLineHeight;
+    final double verticalOffset = -0.5 * preferredLineHeight;
+    return _getTextPositionVertical(offset, verticalOffset);
+  }
+
+  // Returns the TextPosition below the given offset into _plainText.
+  //
+  // If the offset is already on the last line, the given offset will be
+  // returned.
+  TextPosition _getTextPositionBelow(int offset) {
+    // The caret offset gives a location in the upper left hand corner of
+    // the caret so the middle of the line above is a half line above that
+    // point and the line below is 1.5 lines below that point.
+    final double preferredLineHeight = _textPainter.preferredLineHeight;
+    final double verticalOffset = 1.5 * preferredLineHeight;
+    return _getTextPositionVertical(offset, verticalOffset);
+  }
+
+  /// Keeping the current baseOffset fixed, move the extentOffset down by one
```

Comment:
nit: are baseOffset and extentOffset properties can we link them to their API docs by putting them in []?

Reply:
(this is also true for the doc comments below)